File: LeetCode714.cpp
#include<vector>
#include<iostream>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
        int n = prices.size();
        vector<vector<int>> dp(n, vector<int>(2));
        dp[0][0] = 0, dp[0][1] = -prices[0];
        for (int i = 1; i < n; ++i) {
            dp[i][0] = max(dp[i - 1][0], dp[i - 1][1] + prices[i] - fee);
            dp[i][1] = max(dp[i - 1][1], dp[i - 1][0] - prices[i]);
        }
        return dp[n - 1][0];
    }
};

int main(){
    vector<int> prices;
    int n = 0, t = 0, pee;
    cout << "number of nums: ";
    cin >> n;
    for (int i = 1; i <= n; i++){
        cin >> t;
        prices.push_back(t);
    }
    
    cout << "pee: ";
    cin >> pee;
    Solution sol;
    cout << sol.maxProfit(prices, pee);
    return 0;
}
// number of nums: 6
// 1 3 2 8 4 9
// pee: 2
// 8

File: LeetCode714.java
import java.util.Scanner;

public class LeetCode714 {
    public static void main(String[] args) {
        //一维整数数组
        LeetCode714 solution = new LeetCode714();
        Scanner scanner = new Scanner(System.in);
        
        int rows = 0;
        System.out.print("number of nums: ");
        rows = scanner.nextInt();
        int [] prices = new int[rows];
        for (int i = 1; i <= rows; i++){
            prices[i-1] = scanner.nextInt();
        }
        System.out.print("fee: ");
        rows = scanner.nextInt();
        solution.maxProfit(prices, rows);
        scanner.close();
    }
    public int maxProfit(int[] prices, int fee) {
        int n = prices.length;
        int sell = 0, buy = -prices[0];
        for (int i = 1; i < n; ++i) {
            sell = Math.max(sell, buy + prices[i] - fee);
            buy = Math.max(buy, sell - prices[i]);
        }
        return sell;
    }
}
// number of nums: 6
// 1 3 2 8 4 9
// pee: 2
// 8